tests/4.5/target_enter_data/test_target_enter_data_classes_simple.cpp

```diff
@@ -0,0 +1,146 @@
+//===---- test_target_enter_data_classes_simple.c -- target enter data with classes===//
+//
+// OpenMP API Version 4.5 Nov 2015
+//
+//
+// This test checks for interoperability between C++ classes and device offloading through the use of 
+// target enter and exit data clauses. It considers that 4.5 does not support
+// mapping of attributes directly, as the implicit use of the this-> pointer when 
+// using attributes inside the target region is restrictive.
+//
+// The description of the map clause says that the map clause receives a list item:
+//
+// Section 2.15.5.1, page 216, line 17
+//
+// The syntax of the map clause is as follows:
+// map([ [map-type-modifier[,]] map-type : ] list)
+// And the definition of list item is
+//
+// Section 2.1, page 27, line 20
+//
+// A list item is a variable or array section. An extended list item is a list item or a function name.
+//
+// This test creates a class that, during construction, it maps an attribute through helper 
+// variables that remove the direct use to the attributes. And during destruction of the 
+// object it maps the data back to the devices. Additionally, there is a modifier 
+// method that uses values from the class indirectly through the use of helper references
+// finally there is a synchronization  clause that will obtain the values on demand
+//===----------------------------------------------------------------------------------===//
+//
+//
+
+#include <iostream>
+#include <omp.h>
+#include <cassert>
+#include "ompvv.h"
+#include <cmath>
+
+#define N 1000
+
+class Simple {
+private:
+  int *d_array;
+  int size;
+  int sum;
+
+public:
+  // Constructor. Maps the data into the de3vice
+  Simple(int s) : size(s) { 
+    this->sum = 0;
+    this->d_array = new int[size];
+    std::fill(d_array, d_array+size, 0);
+    // Removing the  direct use  of attributes to avoid problems
+    // with 4.5 specifications 
+    int* helper = d_array;
+    int &hs = size;
+    int &hsum = sum;
+#pragma omp target enter data map(to: helper[0:hs]) map(to: hs) map(to:hsum)
```

Comment:
Why take the reference of an `int`?  It's not a heavy-weight object.

And a related question: does this do what you expect?  I.e., does mapping an `int &` map the whole `int`?

Reply:
I believe that some compilers take the whole this-> pointer too serious. I would have to test it again to see if this was the real reason, but I think that using a variable inside that is a member implied a use of the this pointer in the implicit mapping.

This is part of the whole confusion with the this pointer and 4.5

I will try it out without it and see what happens.

Reply:
I think there's a misunderstanding here.

I understand why you're avoiding referring to member symbols directly, that's not what I'm asking.  What I'm asking is:  why did you do `int &hsum = sum;` and not `int hsum = sum;`?

Reply:
I wanted to make sure what gets map is the same memory location. If we map without the reference then it won't correspond to the same location across target regions.  However, since here we are mapping always tofrom it should not make a difference. 

One possible change would be to do:
```C
/// On constructor
int &hsum = sum;
#pragma omp target enter data map(to: hsum)
///...
/// on modify
int &hsum = sum 
#pragma omp target map (alloc:hsum)
/// ...
/// on get values
int &hsum = sum 
#pragma omp target exit data map(from:hsum

```

Reply:
Ahhh, I see.  Thanks for explaining that.